### ISSUE TITLE
v2: API Enhancement: Why is RenderNode part of the Renderer interface?

### ISSUE THREAD

**Issue by client9** [russross/blackfriday#368]
Hello v2 question.

This is the interface for a `Renderer` is:

```
type Renderer interface {
	Render(ast *Node) []byte
	RenderNode(w io.Writer, node *Node, entering bool) WalkStatus
}
```

Why is `RenderNode` part of the interface?  I don't see any direct usage of it.  I do see `Render` called.  It's not clear to me why implementers of  a new renderer need to use this method name or signature (and make it public).   But perhaps I missed something, as `node.Walk` and `RenderNode` are similar.

thanks!

n

**Comment by rtfb** [comment]
Frankly, I don't remember... I hoped `git blame` would help, but apparently I added it with 2f1f0b6b, which doesn't explain much :-/

This commit (0382dab0) provides some insight into what was the intention to separate the methods, but it doesn't explain why I added it to the interface.

**Comment by client9** [comment]
FWIW, I almost have an identity rendered.  I don't use the io.Writer argument since i need to keep my own stack of buffers and I don't have a way of _setting_ the writer in the API.

I think it's save to remove RenderNode this from interface.  You can keep the function and I suspect everything will continue to work.

**Comment by rtfb** [comment]
> FWIW, I almost have an identity rendered.

Neat! Do you want to submit a PR for that as well? I will be travelling for two weeks with limited access to the web and no proper keyboard, so don't be dismayed if I delay to react :-)

Removing `RenderNode` also seems reasonable at the surface. @Ambrevar, @shurcooL, do you have a take on the `RenderNode` issue?

**Comment by client9** [comment]
@rtfb re: identity renderer... it is not exactly an identity renderer but will be easy enough to make it one.     It is in another repo.  When it is closer to being done, I'll file a new ticket and you can review.  I'm happy to donate it to BF.

enjoy your trip!

**Comment by Ambrevar** [comment]
I do use the `RenderNode` function to write custom walkers.
The point of RenderNode is that it returns a WalkStatus which allows for changing the walking path depending on how the current node was traversed.

It might be superfluous though: I'll give it another look soon and let you know.

**Comment by Ambrevar** [comment]
`RenderNode` only renders the node, while `Render` also renders the document header regardless of the node that is passed.
`Render` is a convenience function while `RenderNode` is the underlying function that actually performs the rendering job.
Since the code for rendering the document header is not exported, it is not possible to replace `Render` with `RenderNode` only without rewriting the header -generation code.

I think both functions have to remain part of the interface. There is still room for dicussing the names and the prototypes though.

What about `Render -> RenderDocument` and `RenderNode -> Render`?
(Might not be worth the trouble of confusing user following the transition.)

**Comment by client9** [comment]
Ah thanks for clarification.  I still find having both a bit odd... I'll need to take a look at

> Since the code for rendering the document header is not exported,
> it is not possible to replace Render with RenderNode only without
> rewriting the header -generation code.
>

to help understand your point.


The JSON and XML renderers (that are similar to ASTs sorta) I've used in the past only have a single "RenderNode"type interface  (for instance all the Marshall interfaces in golang, but also in Java, etc)  so I'm trying to understand why markdown is different, and we can make it more similar.

Right now I don't use the Writer arg since the API isn't good enough.  To make it work, I'd need some way of setting the writer that is used since I don't or can't always use the io.Writer that is passed in (I need to use tmp buffers depending on state).  I think there is an assumption that  `Render` must call ast.Walk, and use the function arg ast.Walk requires?

and/or

Render is odd since it returns a []byte, but forces the RenderNode method to use an io.Writer.   i would think Render should take an io.Writer as arg if we want to follow this module.

Something isn't connecting in my head.  I'll try to write up some constructive alternatives later today to help illustrate my concern.

thanks for looking!

n

**Comment by Ambrevar** [comment]
XML/Json do not have document headers.
Besides, Markdown cannot be stream-oriented like those Go libs because it needs handling references which requires a complete parsing of the input.

The `[]byte` returned by `Render` is the final result and should not be modified. Alterations to the rendering should happen before, with `RenderNode`. The latter is on the contrary called many times, with bytes being appended to the `io.Writer`.

**Comment by Ambrevar** [comment]
That being said, we could change the API a bit by replacing `Render` with exported functions, e.g. `Write*Header` and `Write*Footer`, which would not be part of the interface.

In practice, it means the end-user would have to write something like this for every generator:
```
bf.WriteHTMLHeader(&buf, renderer)
ast.Walk(func(node *Node, entering bool) WalkStatus {
	return r.RenderNode(&buf, node, entering)
})
bf.WriteHTMLFooter(&buf, renderer)
```

Short and simple.

Those `Write*` functions are renderer-specific, thus they should not be part of the interface.

It is good practice to keep the interfaces to a minimum, and here there is a possibility to remove one function indeed. What do you think?

**Comment by client9** [comment]
hi @Ambrevar   Looking better, but I need to dig in a bit more on your use case and understand the doc headers/footers more since I'm not currently using them and don't know where they come from in the current API.

To help bridge the gap, I'll tell you my use case:

* I'm not writing out HTML, although I suspect HTML is the most common use case.
* I'm not writing headers or footers (or if I was, it would outside of blackfriday's scope).
* Likewise, I'm not assembling things like tables of contents where, you are correct, I'd need to read the entire AST before generating any output.
* I am (or could if the API wasn't a []byte output) currently streaming top level children (but I do have to read in and parse the entire child before sending).

So I'm trying to figure out how to bridge the gap here. Right now I'm more or less skipping the Render interface, and just using ast.Walk (great!)

Give me a day to take a look and try to come up with something more constructive.

I appreciate your feedback!

n

**Comment by Ambrevar** [comment]
You are correct in all your points, I'm not sure what puzzles you...

**Comment by client9** [comment]
Hi, sorry for the delay.  My confusion was on the HTML header/footer and where those come from and what the API.  I see they are just hardwired in, not API, but one can turn it off or on with flags. 

Taking a look at the HTML renderer, I see it is using a `bytes.Buffer` but is crippled by interface requiring a `io.Writer`.   Allow the render code to use  buffer would eliminate a lot of the casing in the current impl.

This looks better:

```
bf.WriteHTMLHeader(&buf, renderer)
ast.Walk(func(node *Node, entering bool) WalkStatus {
	return r.RenderNode(&buf, node, entering)
})
bf.WriteHTMLFooter(&buf, renderer)
```

So I think we've turned:

```
type Renderer interface {
	Render(ast *Node) []byte
	RenderNode(w io.Writer, node *Node, entering bool) WalkStatus
}
```

into just

```
type Renderer interface {
   Render(ast *Node) []byte
}
```

To make this more generic, I would change this to:

```
type Renderer interface {
   Render(ast *Node, w io.Writer) // error?
}
```

There was a ticket for this actually, that I mistaken suggested was closed:
  https://github.com/russross/blackfriday/issues/120

This would allow streaming output for applications that needed it (for instance our HTML header could be emitted right away).  Also many applications will already have an `io.Writer` (think streaming output to a socket or HTTP response).   Of course the current implementation could keep using a `bytes.Buffer` and emit everything at the end.

discussion welcome!

thanks guys

n

**Comment by Ambrevar** [comment]
No, with my above suggestion, we are removing the need for the `Render` function, so the interface could become

```
type Renderer interface {
	RenderNode(w io.Writer, node *Node, entering bool) WalkStatus
}
```
or, if we rename the function,
```
type Renderer interface {
	Render(w io.Writer, node *Node, entering bool) WalkStatus
}
```
Your "more generic" solution is missing some status...

@rtfb: Didn't you mention somewhere that streaming was not possible because of cross-references?
What about allowing dangling references and linking past references on future input?

**Comment by shurcooL** [comment]
> Removing `RenderNode` also seems reasonable at the surface. @Ambrevar, @shurcooL, do you have a take on the `RenderNode` issue?

I don't have insights on this at the moment. I haven't used v2 or implemented custom renders enough to have an opinion. In general, the fewer methods an interface has, the better, so if it can be removed while preserving functionality, that's the way to go. That's all I got on the matter for now.

**Comment by client9** [comment]
Ah i see some of the confusion is the current `Run` function is `[]byte` in and `[]byte` out, so yah the `io.Writer`  doesn't make much sense right now..  But later we might add a`RunWriter` function (See #120).  so it would be nice to use the `io.Writer` in this interface.

Doing async is hard, and I think @Ambrevar and I are close (hah, maybe), so I'll write up an example so it's more clear.

In my world, this is how the default HTML render would be done.  This is written in the GitHub window so there may be typos.  This might be or not the same thing @Ambrevar mentioned previous (my apologies)


```go
package blackfriday

var htmlHeader := []byte{...} // or existing functions.. this was just for completeness
var htmlFooter := []byte{... }

struct HTMLRender {
   // what you have today
}

// without error handling,  for clarity
func (r HTMLRender) Render(root *node, w io.Writer) error {
    w.Write(htmlHeader)
    buf := bytes.Buffer{} // or do this first
    root.Walk(func(node *Node, entering bool) WalkStatus {
  	return r.renderNode(&buf, node, entering)
    })
    w.Write(buf.Bytes())
    w.Write(htmlFooter)  //  or write to buffer first.
   return nil
}

func (r HTMLRender) renderNode(...) { 
 // you could make this public or not, but not part of the interface.
 // same as what you have now except you can pass in a bytes.Buffer and skip all the casting
}
```

You win since you can skip casting and just use a bytes.Buffer for the default HTML renderer.  That will make the existing code simpler and easier to read (btw, it's quite good already).

I win since I'm not tried to a `[]byte` output.   In my renderer I don't currently use the `io.Writer` passed in `RenderNode` since it's too dumb.  I have to manage intermediate buffers and manage my own state.   But when I have a fully rendered block I can stream it out (with a `RunWriter`).  I can post an example if someone needs to see why this is important.

Everyone wins since there is only one function in the interface.

it's very likely most renderers will do something like this:

```
    buf := bytes.Buffer{} // or do this first
    root.Walk(func(node *Node, entering bool) WalkStatus {
  	return r.renderNode(&buf, node, entering)
    })
    w.Write(buf.Bytes())
```

and use the `type NodeVisitor func(node *Node, entering bool) WalkStatus` interface.

Thats ok since it's already public and defined.

What do you think? 

thanks all for feedback.

n

**Comment by Ambrevar** [comment]
You misunderstood a few points:

- First, the one function we _want_ to keep is `RenderNode`, because it gives us complete flexibility over the rendering process. `Render` is a blackbox. The point of the V2 rewrite is to make the API much more flexible than V1 is.

- In the current state of V2, you will not be able to "stream" blocks of Markdown because references can only be processed over a complete input. Otherwise you'd end up with unresolved references. @rtfb can confirm.

**Comment by client9** [comment]
thanks @Ambrevar  It's clear we have different ideas on interface/api design.  Thats ok.  I'll use the raw AST api.  I'll close this ticket and thanks for your time!  regards, nickg

**Comment by Ambrevar** [comment]
Wait, can you re-open this? We still need to discuss whether we keep 2 functions or just 1 in the interface.

**Comment by rtfb** [comment]
Hey, @client9 and @Ambrevar, thanks for having this discussion and sorry for the delay to get back to it!

First, if there's that much confusion around `Renderer`, it clearly needs improvement or at least documentation. Let's fix that.

Next, regarding streaming. Streaming has two aspects to it: the API interface and the implication that the underlying implementation will actually stream. Unfortunately, it is not possible to construct a fully streaming pipeline when processing markdown. Here's why: the pipeline, in essence, looks like this: `<input markdown document>` -> `parser` -> `AST` -> `renderer` -> `<output document>`. The parser has to grok the entirety of the markdown document and produce the entire AST because it can not know which parts of it the renderer will use. This is, by definition, buffering, not streaming. This is what @Ambrevar referred to.

And that is the reason why I decided that exposing a streaming API would be misleading. Looking at such API would imply the user could stream all the way from a large input document to some output stream in the rendered format, but that would not be true. There would be an in-memory copy at least as large as the source document and potentially larger.

`Renderer` interface is indeed intended to be used with `Run(input, WithRenderer(r))`. It has to expose an interface capable of producing an entire output document. We can change it as following:

``` go
type Renderer interface {
	RenderHeader(w io.Writer, ast *Node)
	RenderFooter(w io.Writer, ast *Node)
	RenderNode(w io.Writer, node *Node, entering bool) WalkStatus
}
```

And finally,

> I'll use the raw AST api

@client9, with the new API this is a perfectly normal approach and might even be recommended for things like you're doing. Let us know how usable is the AST API.

**Comment by rtfb** [comment]
I've fiddled with what was discussed above and pushed #376.

In addition to cleaning up the `Renderer`, this turned out to drop the `OmitContents` flag and I like that a lot. One of the driving factors for v2 was to create an API flexible enough for people to create their niche features for themselves instead of asking us to maintain it. `OmitContents` seems to ~~be~~ have been precisely that.

**Comment by Ambrevar** [comment]
Instead of 2, we now have 3 functions in the interface... Ha! :)

It obviously does not compare:

- The Render/RenderNode overlap is gone. Good.

- The RenderHeader/RenderFooter functions allows for more flexibility: it is now possible to define a new renderer reusing the Footer/Header of other renderers.

Conclusion: I'm all in for that new interface.

Shouldn't we inline writeDocumentHeader?

Why the "WriteString->Write" rename?  Is it actually faster?  It's a tiny bit harder to read and costs 2 chars on every call...
Anyways, it would have been cleaner in a seperate commit I think.

**Comment by rtfb** [comment]
> Shouldn't we inline writeDocumentHeader?

Yeah, I considered, but decided against it because `HTMLRenderer.RenderHeader` does exactly two things and without inlining this is more clearly visible.

> Why the "WriteString->Write" rename?

Purely due to `bytes.Buffer` -> `io.Writer` change. The latter doesn't have a `WriteString`.

**Comment by Ambrevar** [comment]
Acknowleged, I had missed the `bytes.Buffer`->`io.Writer` transition.

**Comment by rtfb** [comment]
BTW, changed that to `io.WriteString(w, "string")`, like @shurcooL suggested.
